From 27162cd64ecf52c2b61073c6021d39fe8e5c2718 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Jonah=20Br=C3=BCchert?= <jbb@kaidan.im>
Date: Mon, 24 Feb 2025 13:03:49 +0100
Subject: [PATCH 1/3] Fix loading albums without id

After all this is a bug in ytmusicapi, but we can handle this more
gracefully
---
 src/albummodel.cpp             | 8 ++++++--
 src/contents/ui/AlbumPage.qml  | 7 ++++++-
 src/contents/ui/ListHeader.qml | 2 ++
 src/example/main.cpp           | 2 +-
 src/ytmusic.cpp                | 2 +-
 src/ytmusic.h                  | 2 +-
 6 files changed, 17 insertions(+), 6 deletions(-)

diff --git a/src/albummodel.cpp b/src/albummodel.cpp
index d4c070ea..9febd4e2 100644
--- a/src/albummodel.cpp
+++ b/src/albummodel.cpp
@@ -106,12 +106,16 @@ QUrl AlbumModel::thumbnailUrl() const
 
 QString AlbumModel::playlistId() const
 {
-    return QString::fromStdString(m_album.audio_playlist_id);
+    return QString::fromStdString(m_album.audio_playlist_id.value_or(std::string()));
 }
 
 QUrl AlbumModel::webUrl() const
 {
-    return QUrl(YTMUSIC_WEB_BASE_URL % "/playlist?list=" % QString::fromStdString(m_album.audio_playlist_id));
+    if (m_album.audio_playlist_id) {
+        return QUrl(YTMUSIC_WEB_BASE_URL % "/playlist?list=" % QString::fromStdString(*m_album.audio_playlist_id));
+    }
+
+    return {};
 }
 
 const album::Album &AlbumModel::album() const
diff --git a/src/contents/ui/AlbumPage.qml b/src/contents/ui/AlbumPage.qml
index 44736e5c..fc30cec8 100644
--- a/src/contents/ui/AlbumPage.qml
+++ b/src/contents/ui/AlbumPage.qml
@@ -20,7 +20,7 @@ Kirigami.ScrollablePage {
         id: action
         visible: false
         property bool shown
-        shown: !root.flickable.atYBeginning
+        shown: !root.flickable.atYBeginning && albumModel.playlistId
         onShownChanged:
             if(shown){
                 visible = true
@@ -69,6 +69,7 @@ Kirigami.ScrollablePage {
                 Kirigami.Action {
                     icon.name: "media-playback-start"
                     text: i18n("Play")
+                    visible: albumModel.playlistId
                     onTriggered: {
                         applicationWindow().playPlaylist(albumModel.playlistId)
                     }
@@ -76,6 +77,7 @@ Kirigami.ScrollablePage {
                 Kirigami.Action {
                     icon.name: "media-playlist-shuffle"
                     text: i18n("Shuffle")
+                    visible: albumModel.playlistId
                     onTriggered: {
                         applicationWindow().playShufflePlaylist(albumModel.playlistId)
                     }
@@ -86,17 +88,20 @@ Kirigami.ScrollablePage {
                     text: i18n("Append to queue")
                     icon.name: "media-playlist-append"
                     onTriggered: UserPlaylistModel.appendAlbum(albumModel)
+                    visible: albumModel.playlistId
                 },
                 Kirigami.Action {
                     text: i18n("Open in Browser")
                     icon.name: "internet-services"
                     onTriggered: Qt.openUrlExternally(albumModel.webUrl)
+                    visible: albumModel.playlistId
                 },
                 Kirigami.Action {
                     text: i18n("Share")
                     icon.name: "emblem-shared-symbolic"
                     onTriggered:{
                         openShareMenu(albumModel.title, albumModel.webUrl)
+                        visible: albumModel.playlistId
                     }
                 }
 
diff --git a/src/contents/ui/ListHeader.qml b/src/contents/ui/ListHeader.qml
index 8689c154..d65595c7 100644
--- a/src/contents/ui/ListHeader.qml
+++ b/src/contents/ui/ListHeader.qml
@@ -131,6 +131,7 @@ Item {
                             required property var modelData
                             icon.name: modelData.icon.name
                             text: modelData.text
+                            visible: modelData.visible
                             onTriggered: modelData.triggered()
                         }
                     }
@@ -151,6 +152,7 @@ Item {
                                 Layout.fillWidth: true
                                 text: modelData.text
                                 icon.name: modelData.icon.name
+                                visible: modelData.visible
                                 onClicked: {
                                     modelData.triggered()
                                     overflowDrawer.close()
diff --git a/src/example/main.cpp b/src/example/main.cpp
index 3fa1f5c2..f5ede345 100644
--- a/src/example/main.cpp
+++ b/src/example/main.cpp
@@ -8,7 +8,7 @@
 
 int main() {
     const auto ytm = YTMusic();
-    const auto results = ytm.search("Systemabsturz");
+    const auto results = ytm.search("Kargo");
 
     std::cout << "Found " << results.size() << " results." << std::endl;
 
diff --git a/src/ytmusic.cpp b/src/ytmusic.cpp
index fd30227d..15414898 100644
--- a/src/ytmusic.cpp
+++ b/src/ytmusic.cpp
@@ -427,7 +427,7 @@ album::Album YTMusic::get_album(const std::string &browseId) const
         album["title"].cast<std::string>(),
         album["trackCount"].cast<int>(),
         album["duration"].cast<std::string>(),
-        album["audioPlaylistId"].cast<std::string>(),
+        optional_key<std::string>(album, "audioPlaylistId"),
         optional_key<std::string>(album, "year"),
         optional_key<std::string>(album, "description"),
         extract_py_list<meta::Thumbnail>(album["thumbnails"]),
diff --git a/src/ytmusic.h b/src/ytmusic.h
index 0eb051f2..7b4cc738 100644
--- a/src/ytmusic.h
+++ b/src/ytmusic.h
@@ -177,7 +177,7 @@ namespace album {
         std::string title;
         int track_count;
         std::string duration;
-        std::string audio_playlist_id;
+        std::optional<std::string> audio_playlist_id;
         std::optional<std::string> year;
         std::optional<std::string> description;
         std::vector<meta::Thumbnail> thumbnails;
-- 
2.48.1

